File: benchmarks/fixtures/webapp_ts/src/auth/service.ts
/** Authentication service with class hierarchy. */

import { getLogger } from '../utils/helpers';
import { validateToken, generateToken, TokenError } from './tokens';

const logger = getLogger("auth.service");

/** Base service providing shared functionality. */
export class BaseService {
    protected serviceName: string;
    protected initialized: boolean;

    constructor(serviceName: string = "base") {
        this.serviceName = serviceName;
        this.initialized = false;
    }

    /** Initialize the service. */
    initialize(): void {
        this.initialized = true;
        logger.info(`${this.serviceName} initialized`);
    }

    /** Shut down the service. */
    shutdown(): void {
        this.initialized = false;
        logger.info(`${this.serviceName} shut down`);
    }

    /** Check service health. */
    healthCheck(): { service: string; status: string } {
        return {
            service: this.serviceName,
            status: this.initialized ? "healthy" : "not_initialized",
        };
    }

    /** Ensure service is initialized. */
    protected requireInitialized(): void {
        if (!this.initialized) {
            throw new Error(`${this.serviceName} not initialized`);
        }
    }
}

/** User record interface. */
export interface User {
    id: string;
    email: string;
    name: string;
    role: string;
    passwordHash: string;
}

/** Authentication service handling login, logout, and user management. */
export class AuthService extends BaseService {
    private users: Map<string, User>;

    constructor() {
        super("auth");
        this.users = new Map();
    }

    /** Authenticate user and return token. */
    async login(email: string, password: string): Promise<string | null> {
        this.requireInitialized();
        logger.info(`Login attempt for ${email}`);
        const user = this._findUser(email);
        if (!user) {
            logger.warn(`User not found: ${email}`);
            return null;
        }
        // Verify password (simulated)
        if (user.passwordHash !== password) {
            return null;
        }
        const token = generateToken(user);
        logger.info(`Login successful for ${email}`);
        return token;
    }

    /** Log out a user by invalidating their token. */
    async logout(token: string): Promise<boolean> {
        this.requireInitialized();
        logger.info("Processing logout");
        try {
            validateToken(token);
            return true;
        } catch (e) {
            if (e instanceof TokenError) {
                logger.warn("Invalid token on logout");
            }
            return false;
        }
    }

    /** Get the current user from a token. */
    async getCurrentUser(token: string): Promise<User | null> {
        this.requireInitialized();
        try {
            const claims = validateToken(token);
            return this._findUser(claims.email) ?? null;
        } catch {
            return null;
        }
    }

    /** Change user password. */
    async changePassword(userId: string, oldPassword: string): Promise<boolean> {
        this.requireInitialized();
        logger.info(`Password change for ${userId}`);
        // Would verify old password and set new one
        return true;
    }

    /** Find a user by email. */
    _findUser(email: string): User | undefined {
        for (const user of this.users.values()) {
            if (user.email === email) {
                return user;
            }
        }
        return undefined;
    }
}

/** Admin service extending AuthService. */
export class AdminService extends AuthService {
    constructor() {
        super();
        this.serviceName = "admin";
    }

    /** Impersonate a user (admin only). */
    async impersonate(adminToken: string, targetUserId: string): Promise<string | null> {
        this.requireInitialized();
        const admin = await this.getCurrentUser(adminToken);
        if (!admin || admin.role !== "admin") {
            return null;
        }
        logger.info(`Admin impersonating user ${targetUserId}`);
        return generateToken({ id: targetUserId, email: "", role: "user" });
    }

    /** List all users (admin only). */
    async listAllUsers(adminToken: string): Promise<User[]> {
        this.requireInitialized();
        const admin = await this.getCurrentUser(adminToken);
        if (!admin || admin.role !== "admin") {
            return [];
        }
        return [];
    }
}

File: benchmarks/fixtures/webapp_go/internal/auth/service.go
package auth

import (
    "fmt"

    "webapp_go/pkg/logger"
)

var serviceLog = logger.GetLogger("auth.service")

// User represents an authenticated user.
type User struct {
    ID       string
    Email    string
    Password string
    Name     string
    Role     string
    Active   bool
}

// AuthProvider defines the interface for authentication providers.
type AuthProvider interface {
    Login(email, password string) (string, error)
    Logout(token string) error
}

// BaseService provides common service functionality.
type BaseService struct {
    Name    string
    Version string
}

// Initialize sets up the base service.
func (b *BaseService) Initialize() {
    serviceLog.Info("Initializing service: %s v%s", b.Name, b.Version)
}

// AuthService handles user authentication.
type AuthService struct {
    BaseService
    Users map[string]*User
}

// NewAuthService creates a new authentication service.
func NewAuthService() *AuthService {
    svc := &AuthService{
        BaseService: BaseService{Name: "auth", Version: "1.0"},
        Users:       make(map[string]*User),
    }
    svc.Initialize()
    serviceLog.Info("AuthService created")
    return svc
}

// Login authenticates a user and returns a token.
func (s *AuthService) Login(email, password string) (string, error) {
    serviceLog.Info("Login attempt for: %s", email)
    user, ok := s.Users[email]
    if !ok {
        serviceLog.Warn("User not found: %s", email)
        return "", fmt.Errorf("user not found: %s", email)
    }
    if user.Password != password {
        serviceLog.Warn("Invalid password for: %s", email)
        return "", fmt.Errorf("invalid credentials")
    }
    if !user.Active {
        serviceLog.Warn("Inactive user: %s", email)
        return "", fmt.Errorf("account disabled")
    }
    token := GenerateToken(*user)
    serviceLog.Info("Login successful for: %s", email)
    return token, nil
}

// Logout invalidates a user's token.
func (s *AuthService) Logout(token string) error {
    serviceLog.Info("Logout request")
    return RevokeToken(token)
}

// GetCurrentUser returns the user associated with a token.
func (s *AuthService) GetCurrentUser(token string) (*User, error) {
    serviceLog.Info("Getting current user from token")
    claims, err := ValidateToken(token)
    if err != nil {
        serviceLog.Error("Invalid token: %v", err)
        return nil, err
    }
    user := &User{
        ID:    claims.UserID,
        Email: claims.Email,
        Role:  claims.Role,
    }
    serviceLog.Info("Current user: %s", user.Email)
    return user, nil
}

// AdminService extends AuthService with admin-specific functionality.
type AdminService struct {
    AuthService
    AdminUsers []string
}

// NewAdminService creates a new admin service.
func NewAdminService() *AdminService {
    svc := &AdminService{
        AuthService: *NewAuthService(),
        AdminUsers:  []string{},
    }
    serviceLog.Info("AdminService created")
    return svc
}

// IsAdmin checks whether a user has admin privileges.
func (a *AdminService) IsAdmin(userID string) bool {
    serviceLog.Debug("Checking admin status for: %s", userID)
    for _, id := range a.AdminUsers {
        if id == userID {
            return true
        }
    }
    return false
}

// PromoteToAdmin grants admin privileges to a user.
func (a *AdminService) PromoteToAdmin(userID string) {
    serviceLog.Info("Promoting user to admin: %s", userID)
    a.AdminUsers = append(a.AdminUsers, userID)
}
